parachain/runtime/common/src/evm/multi_currency.rs

```diff
@@ -0,0 +1,426 @@
+//! ERC20 compatible EVM precompile(s) for interacting with currencies.
+
+use evm_utils::*;
+use fp_evm::{ExitError, PrecompileFailure};
+use frame_support::{
+    dispatch::{GetDispatchInfo, PostDispatchInfo},
+    pallet_prelude::MaxEncodedLen,
+};
+use orml_traits::MultiCurrency;
+use pallet_evm::{AddressMapping, GasWeightMapping, IsPrecompileResult, PrecompileHandle, PrecompileResult};
+use primitives::{is_currency_precompile, CurrencyId, CurrencyInfo, ForeignAssetId, LpToken, StablePoolId};
+use sp_core::{Get, H160, U256};
+use sp_runtime::traits::{Dispatchable, StaticLookup};
+use sp_std::{marker::PhantomData, prelude::*};
+use xcm::VersionedMultiLocation;
+
+pub trait CurrencyApis:
+    orml_asset_registry::Config<AssetId = ForeignAssetId>
+    + loans::Config<CurrencyId = CurrencyId>
+    + dex_stable::Config<CurrencyId = CurrencyId, PoolId = StablePoolId>
+{
+}
+
+impl<Api> CurrencyApis for Api where
+    Api: orml_asset_registry::Config<AssetId = ForeignAssetId>
+        + loans::Config<CurrencyId = CurrencyId>
+        + dex_stable::Config<CurrencyId = CurrencyId, PoolId = StablePoolId>
+{
+}
+
+pub trait RuntimeCurrencyInfo {
+    fn max_db_read<T: CurrencyApis>() -> usize;
+    fn name<T: CurrencyApis>(&self) -> Option<Vec<u8>>;
+    fn symbol<T: CurrencyApis>(&self) -> Option<Vec<u8>>;
+    fn decimals<T: CurrencyApis>(&self) -> Option<u32>;
+}
+
+struct ForeignAssetInfo(ForeignAssetId);
+
+impl RuntimeCurrencyInfo for ForeignAssetInfo {
+    // NOTE: `name`, `symbol` and `coingecko_id` are not bounded
+    // estimate uses 32 bytes each
```

Comment:
We can fix this once https://github.com/open-web3-stack/open-runtime-module-library/pull/949 is merged